### PR TITLE
Fixes #RHIROS-1176, #RHIROS-1178 - suggested_instance_types/<instance_type> api endpoint

### PR DESCRIPTION
**Pull request by upadhyeammit**
The suggested_instance_types api endpoint accepting instance_type with filtering defined on display_name, state and OS

## PR Title :boom:

Please title this PR with a summary of the change, along with the JIRA card number.

Suggested formats: 

1. Fixes/Refs #RHIROS-XXX - Title
2. RHIROS-XXX Title 

Feel free to remove this section from PR description once done.

## Why do we need this change? :thought_balloon:

Please include the __context of this change__ here.

## Documentation update? :memo:

- [ ] Yes
- [ ] No

## Security Checklist :lock:

Upon raising this PR please go through [RedHatInsights/secure-coding-checklist](https://github.com/RedHatInsights/secure-coding-checklist)

## :guardsman: Checklist :dart:

- [ ] Bugfix
- [ ] New Feature
- [ ] Refactor
- [ ] Unittests Added
- [ ] DRY code
- [ ] Dependency Added
- [ ] DB Migration Added

## Additional :mega:

Feel free to add any other relevant details such as __links, notes, screenshots__, here.

### CODE REVIEW

File: ros/api/v1/suggested_instance_types.py [hunk 1]
```diff
@@ -96,3 +109,104 @@ def build_sort_expression(self, order_how, order_method):
...
+    @marshal_with(output)
+    def get(self, instance_type):
+        limit = limit_value()
+        offset = offset_value()
+        org_id = org_id_from_identity_header(request)
+        system_query = group_filtered_query(system_ids_by_org_id(org_id))
+        query = db.session.query(
+            System.fqdn,
+            System.display_name,
+            func.concat(System.operating_system["name"].astext, ' ', System.operating_system['major'], '.',
```
Comment (r14chandra) [inline ros/api/v1/suggested_instance_types.py:166]: We have [this](https://github.com/RedHatInsights/ros-backend/blob/main/ros/lib/models.py#L90) function already to deserialize OS data. We can use that.

File: ros/api/v1/suggested_instance_types.py [hunk 2]
```diff
@@ -96,3 +109,104 @@ def build_sort_expression(self, order_how, order_method):
...
+
+    @staticmethod
+    def build_instance_details_filters():
+        filters = []
+        if filter_display_name := request.args.get('display_name'):
+            filters.append(System.display_name.ilike(f'%{filter_display_name}'))
+        if filter_state := request.args.get('state'):
+            filters.append(System.state.ilike(f'%{filter_state}'))
+        if filter_os := request.args.get('os'):
+            filters.append(func.concat(System.operating_system["name"].astext, ' ',
```
Comment (r14chandra) [inline ros/api/v1/suggested_instance_types.py:209]: Same here. We can use the custom deserialize method.

**Review by r14chandra — CHANGES_REQUESTED** [review]

File: ros/api/v1/suggested_instance_types.py (hunk 1 above)
Comment (upadhyeammit) [reply ros/api/v1/suggested_instance_types.py:166]: While trying to use the `deserialize_host_os_data` function I faced below challenges,
1. Its the property of System's model
2. If I want to use it then in query I would need to select all columns of System's model(actually call whole model) even though I don't need all of them

So I have below thoughts,
a. We retrieve all columns from tables and use the function, which I think we should not do?
b. Rewrite the `deserialize_host_os_data` but not as property, this means we would also change existing code which depends on this method.

Let me know your thoughts

Comment (r14chandra) [reply ros/api/v1/suggested_instance_types.py:157]: The second one is better. But I don't know much changes are required in the existing code. We can target it in separate PR or in the same. What do you think?

Comment (upadhyeammit) [reply ros/api/v1/suggested_instance_types.py:157]: As per the discussion we would like to take route b, which would need more code change. Considering this I would have separate pr for that refactoring

**Review by r14chandra — APPROVED** [review]

**Comment by r14chandra** [comment]
@yash2189 Could test the PR before we merge?

**Comment by yash2189** [comment]
Tests are successful for this PR and observations have been added over respective Jira card.

**Comment by upadhyeammit** [comment]
Thanks a lot for reviews and testing! Merging this now!

**Pull request merged**